### PR TITLE
OIDC provider ActiveState

### PR DESCRIPTION
**Pull request by th3coop**
1 or 2 PRs to add ActiveState as an OIDC provider.  This PR adds the backend features:
- DB migration
- associated modeling
- tests
- `disallow-activestate-oidc` admin flag is `True`

As the OIDC code seems to be evolving, I explicitly avoided making type changes outside the new code as I was getting a lot of merge conflicts otherwise. 

ActiveState Internal work tracking ID: DS-1723

### CODE REVIEW

**Comment by th3coop** [comment]
Fixing test coverage.  Took longer that I'd care to admit to figure out how to view the coverage report...:D

**Comment by th3coop** [comment]
Self reviewing WIP PR before requesting a review, FYI.

File: tests/unit/oidc/models/test_activestate.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+class TestActiveStatePublisher:
+    def test_publisher_name(self):
+        publisher = ActiveStatePublisher()
+
+        assert publisher.publisher_name == "ActiveState"
+
+    def test_publisher_url(self):
+        publisher = ActiveStatePublisher()
+
+        assert publisher.publisher_url == "https://platform.activestate.com/oauth/oidc"
```
Comment (th3coop) [inline tests/unit/oidc/models/test_activestate.py:81]: I'm not clear on what this field should be.  Is this meant to match the ISSUER?  Or can this be anything we like?  If it's the latter, I'll change this to `https://platform.activestate.com/`.  If it's the former, it would be more clear field had a different name, indicative of it's meaning.

Comment (di) [reply tests/unit/oidc/models/test_activestate.py:81]: This is not the issuer, it's a URL that uniquely identifies the publisher. E.g. for GitHub, this is a link to a repository, possibly also including a specific commit SHA if one is available for the identity used: https://github.com/pypi/warehouse/blob/5a0289e2d9df46a1e3d823a6c00b9aa74bd03417/warehouse/oidc/models/github.py#L201-L207

Comment (th3coop) [reply tests/unit/oidc/models/test_activestate.py]: Ahhhh, so this is a "publisher" as in the user that owns the code is publishing it to PyPI, not the OIDC publisher/provider.  Ok.  I know what to put here now.

**Comment by th3coop** [comment]
Ok, I think this PR is where I want it.  It's ready for review. Worth noting again (I noted it in the Description as well) that I've disabled the admin flag for ActiveState OIDC for now.  I'm not 100% sure that's proper but seemed prudent.

**Comment by th3coop** [comment]
Sorry for the flip flop, making the PR a Draft again.

I want to run my intended plan by y'all before getting a review:
I want to avoid merging this work into `main` before we ("we" being ActiveState) have had a chance to test it with a real world use case.  That "real world use case" is an internal service that is planned and work will start on it soon.  Until that's done, or at least got it to a point where we can test the flow between PyPI and the service that will publish wheel, I'd like to keep this work in a branch.

I'll be starting the front end PR for this today as well.

Does this plan sound ok?  I'll take responsibility for keeping the languishing PRs up to date with pypi/warehouse:main, of course.

If an open PR hanging in the repo isn't desirable I can of course close the PR and create a new one later when we're ready to proceed.

Any preference or thoughts on how you'd like me to proceed?

**Comment by th3coop** [comment]
Thank you kindly for the quick reply :D

**Comment by th3coop** [comment]
@di.  I rebased this branch off of https://github.com/pypi/warehouse/pull/15148. Everything seems to be all spiffy.  I'll watch Woodruffw's PR incase I should rebase again to pull in any changes that might come up there but right now, it should be good to go for review.

File: warehouse/migrations/versions/bd560a14a75c_.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# limitations under the License.
+"""
+empty message
+
+Revision ID: bd560a14a75c
+Revises: ('9a0ed2044b53', 'e6a1cca38664')
+Create Date: 2024-01-06 00:30:27.963411
+"""
+
+revision = "bd560a14a75c"
```
Comment (di) [inline warehouse/migrations/versions/bd560a14a75c_.py:20]: This seems like a no-op migration

**Comment by di** [comment]
Ok @th3coop, if you can resolve the conflicts here (hopefully for the last time!) I think this will be ready for review. Please take a look at #15144 to see how we're now parametrizing the tests.

File: warehouse/migrations/versions/bd560a14a75c_.py (hunk 2 above)
Comment (th3coop) [reply warehouse/migrations/versions/bd560a14a75c_.py:20]: This was the result of running `docker compose run --rm web python -m warehouse db merge heads`, which results in `alembic merg heads` being run.  It was the recommended way of handling an error I was getting: 
```
Only a single head is supported. The script directory has multiple heads (due branching), which must be resolved by manually editing the revision files to form a linear sequence. 
Run `alembic branches` to see the divergence(s)
```
That was a result of me merging in an upstream change that included a migration, so my migration and that other migration "revised" the same migration, so essentially two branches in the migration graph.  This migration merges things back together.

I am not an alembic pro.  I've gathered that from reading briefly the docs around this command and reviewing related SO threads.  I also checked the other migrations and found that there are 2 other cases where this was done.

I'd be happy to revert my migration and reapply it to keep the graph linear though.  Let me know what you think.

File: tests/unit/oidc/test_views.py [hunk 3]
```diff
@@ -546,7 +544,7 @@ def find_service(iface, **kw):
     assert macaroon_service.create_macaroon.calls == [
         pretend.call(
             "fakedomain",
-            f"OpenID token: fakespecifier ({datetime.fromtimestamp(0).isoformat()})",
+            f"OpenID token: example.yml ({datetime.fromtimestamp(0).isoformat()})",
```
Comment (di) [inline tests/unit/oidc/test_views.py:547]: Is this change necessary?

File: warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+#
+#       If you need to increase this timeout for a migration, you can do so
+#       by adding:
+#
+#           op.execute("SET statement_timeout = 5000")
+#           op.execute("SET lock_timeout = 4000")
+#
+#       To whatever values are reasonable for this migration as part of your
+#       migration.
+
```
Comment (di) [inline warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py:48]:
```suggestion
```

File: warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (di) [inline warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py:51]:
```suggestion
```

File: warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        ),
+        sa.PrimaryKeyConstraint("id"),
+        sa.UniqueConstraint(
+            "organization",
+            "activestate_project_name",
+            "actor_id",
+            name="_pending_activestate_oidc_publisher_uc",
+        ),
+    )
+    # ### end Alembic commands ###
```
Comment (di) [inline warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py:92]:
```suggestion
```

File: warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            'Disallow the ActiveState OIDC provider',
+            TRUE,
+            FALSE
+        )
+        """
+    )
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (di) [inline warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py:108]:
```suggestion
```

File: warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    op.drop_table("pending_activestate_oidc_publishers")
+    op.drop_table("activestate_oidc_publishers")
+    # ### end Alembic commands ###
```
Comment (di) [inline warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py:111]:
```suggestion
```

File: warehouse/oidc/models/_core.py [hunk 9]
```diff
@@ -140,7 +140,6 @@ def lookup_by_claims(cls, session, signed_claims: SignedClaims):
                 # We might not build a query if we know the claim set can't
                 # satisfy it. If that's the case, then we skip.
                 continue
-
```
Comment (di) [inline warehouse/oidc/models/_core.py:143]: Unnecessary change here.

File: warehouse/oidc/models/activestate.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        "project": oidccore.check_claim_binary(str.__eq__),
+        "actor_id": oidccore.check_claim_binary(str.__eq__),
+        # This is the name of the builder in the ActiveState Platform that
+        # publishes things to PyPI.
+        "builder": oidccore.check_claim_invariant("pypi-publisher"),
+    }
+
+    __optional_verifiable_claims__: dict[str, CheckClaimCallable[Any]] = {
+        # Should we check that the pypi projectname matches this ingredient
+        # name?
```
Comment (di) [inline warehouse/oidc/models/activestate.py:92]: What is an "ingredient" for an ActiveState project? Is it always the PyPI project name?

File: warehouse/oidc/models/activestate.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    Common functionality for both pending and concrete ActiveState OIDC publishers.
+    """
+
+    organization = mapped_column(String, nullable=False)
+    activestate_project_name = mapped_column(String, nullable=False)
+    actor = mapped_column(String, nullable=False)
+    # 'actor' (The ActiveState platform username) is obstained from the user
+    # while configuring the publisher We'll make an api call to ActiveState to
+    # get the 'actor_id'
+    actor_id = mapped_column(String, nullable=False)
```
Comment (di) [inline warehouse/oidc/models/activestate.py:77]: Do we need to do the same thing for `organization` that we do for `actor`/`actor_id` here? (i.e., get the `organization_id` and compare against that instead).

**Comment by th3coop** [comment]
@di  looks like the changes in #15144 will effect my other PR, #15168, which adds the ActiveState form.  I'll get that and my other branch updated meow.

File: warehouse/oidc/models/activestate.py [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+        __lookup_all__,
+    ]
+
+    @property
+    def sub(self) -> str:
+        return f"org:{self.organization}:project:{self.activestate_project_name}"
+
+    @property
+    def builder(self) -> str:
+        return "pypi-publisher"
```
Comment (di) [inline warehouse/oidc/models/activestate.py:123]: Is this necessary if we're verifying it as an invariant?

File: warehouse/oidc/models/activestate.py [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    @property
+    def project(self) -> str:
+        return self.activestate_project_name
+
+    @property
+    def activestate_project_path(self) -> str:
+        return f"{self.organization}/{self.activestate_project_name}"
+
+    def publisher_url(self, claims: SignedClaims | None = None) -> str:
+        return urllib.parse.urljoin(ACTIVESTATE_URL, self.activestate_project_path)
```
Comment (di) [inline warehouse/oidc/models/activestate.py:138]:
```suggestion
    def publisher_url(self, claims: SignedClaims | None = None) -> str:
        return urllib.parse.urljoin(
            ACTIVESTATE_URL,
            f"{self.organization}/{self.activestate_project_name}",
        )
```
The `activestate_project_path` property doesn't seem to be used anywhere else.

File: warehouse/oidc/models/activestate.py [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+
+class ActiveStatePublisher(ActiveStatePublisherMixin, OIDCPublisher):
+    __tablename__ = "activestate_oidc_publishers"
+    __mapper_args__ = {"polymorphic_identity": "activestate_oidc_publishers"}
+    __table_args__ = (
+        UniqueConstraint(
+            "organization",
+            "activestate_project_name",
+            # This field is NOT populated from the form but from
```
Comment (di) [inline warehouse/oidc/models/activestate.py:151]: This comment is unfinished.

File: warehouse/oidc/models/github.py [hunk 15]
```diff
@@ -269,7 +269,7 @@ def reify(self, session):
         Returns a `GitHubPublisher` for this `PendingGitHubPublisher`,
         deleting the `PendingGitHubPublisher` in the process.
         """
-
+        print("Github.reify")
```
Comment (di) [inline warehouse/oidc/models/github.py:272]:
```suggestion

```

File: warehouse/oidc/models/github.py [hunk 16]
```diff
@@ -280,6 +280,7 @@ def reify(self, session):
             )
             .one_or_none()
         )
+        print(f"Github.reify: {maybe_publisher}")
```
Comment (di) [inline warehouse/oidc/models/github.py:283]:
```suggestion
```

File: warehouse/oidc/models/github.py [hunk 17]
```diff
@@ -288,6 +289,8 @@ def reify(self, session):
             workflow_filename=self.workflow_filename,
             environment=self.environment,
         )
+        print(f"Github.reify: {publisher}")
```
Comment (di) [inline warehouse/oidc/models/github.py:292]:
```suggestion
```

File: warehouse/oidc/models/github.py [hunk 18]
```diff
@@ -288,6 +289,8 @@ def reify(self, session):
...
 
         session.delete(self)
+        print("Github.reify: deleted successfully?")
```
Comment (di) [inline warehouse/oidc/models/github.py:295]:
```suggestion
```

File: warehouse/oidc/models/google.py [hunk 19]
```diff
@@ -9,7 +9,6 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
```
Comment (di) [inline warehouse/oidc/models/google.py:12]: Unrelated change.

File: warehouse/oidc/utils.py [hunk 20]
```diff
@@ -65,7 +80,6 @@ def find_publisher_by_issuer(session, issuer_url, signed_claims, *, pending=Fals
         # This indicates a logic error, since we shouldn't have verified
         # claims for an issuer that we don't recognize and support.
         raise InvalidPublisherError(f"Issuer {issuer_url!r} is unsupported")
-
```
Comment (di) [inline warehouse/oidc/utils.py:68]: Unnecessary change.

File: warehouse/oidc/views.py [hunk 21]
```diff
@@ -241,7 +241,6 @@ def mint_token(
             ],
             request=request,
         )
-
```
Comment (di) [inline warehouse/oidc/views.py:244]: Unnecessary change.

File: warehouse/migrations/versions/bd560a14a75c_.py (hunk 2 above)
Comment (di) [reply warehouse/migrations/versions/bd560a14a75c_.py:20]: Ah, I missed that, the preferable thing would be to either remove & regenerate the migration, or just update the "down revision" on the latest migration. I made suggestions which should remove the need for the no-op migration.

**Comment by th3coop** [comment]
Oof.  Some silly mistakes in there.  Sorry about that @di.  I'll make sure the next PRs are more up to snuff.

**Comment by th3coop** [comment]
Thanks very much for the review though!  Working on changes now.

File: warehouse/oidc/utils.py [hunk 22]
```diff
@@ -20,32 +20,47 @@
...
-from warehouse.oidc.models._core import OIDCPublisherMixin
 
 GITHUB_OIDC_ISSUER_URL = "https://token.actions.githubusercontent.com"
 GOOGLE_OIDC_ISSUER_URL = "https://accounts.google.com"
+ACTIVESTATE_OIDC_ISSUER_URL = "https://platform.activestate.com/api/v1/oauth/oidc"
 
 OIDC_ISSUER_SERVICE_NAMES = {
     GITHUB_OIDC_ISSUER_URL: "github",
     GOOGLE_OIDC_ISSUER_URL: "google",
+    ACTIVESTATE_OIDC_ISSUER_URL: "google",
```
[the remark this answers is not in the document]
Comment (th3coop) [reply warehouse/oidc/utils.py:40]: 🤕

File: warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Add ActiveState OIDC publisher
+
+Revision ID: 9a0ed2044b53
+Revises: 4297620f7b41
```
[the remark this answers is not in the document]
Comment (th3coop) [reply warehouse/migrations/versions/9a0ed2044b53_add_activestate_oidc_publisher.py:16]: Done locally.

File: tests/unit/oidc/test_views.py [hunk 24]
```diff
@@ -419,13 +419,13 @@ def test_mint_token_from_pending_trusted_publisher_invalidates_others(
 
     user = UserFactory.create()
     pending_publisher = PendingGitHubPublisherFactory.create(
-        project_name="does-not-exist",
+        project_name="DOES-NOT-EXIST",
```
[the remark this answers is not in the document]
Comment (th3coop) [reply tests/unit/oidc/test_views.py:422]: Not necessary.  I changed it because it was identical to one of the invalidated projects in the code below this.  It seemed like the code was partially testing project name normalization so made the project name yet another variation that normalization would cause to match.  Will revert though to reduce noise.

File: tests/unit/oidc/test_views.py [hunk 25]
```diff
@@ -500,8 +500,6 @@ def test_mint_token_no_pending_publisher_ok(
     publisher = GitHubPublisherFactory()
     monkeypatch.setattr(publisher.__class__, "projects", [project])
     publisher.publisher_url = pretend.call_recorder(lambda **kw: "https://fake/url")
-    # NOTE: Can't set __str__ using pretend.stub()
-    monkeypatch.setattr(publisher.__class__, "__str__", lambda s: "fakespecifier")
```
[the remark this answers is not in the document]
Comment (th3coop) [reply tests/unit/oidc/test_views.py:504]: It's been a while since I made this change.  If I bring this back the assertion at line 544 will fail, specifically [line 547](https://github.com/ActiveState/warehouse/blob/OIDC-provider-ActiveState/tests/unit/oidc/test_views.py#L547).  I think I removed it because I thought it was testing GitHub specifically, given the claims being passed in, so what the publishers `__str__` function actually returns made sense to test.  Having said that, that would mean I should have removed the stub of `publisher_url`.  

I might have missed the intention of this test here.  Still want me to bring this back and fix change the assertion at line 547?

File: tests/unit/oidc/test_views.py (hunk 3 above)
Comment (th3coop) [reply tests/unit/oidc/test_views.py:547]: Related https://github.com/pypi/warehouse/pull/13980#discussion_r1448020448

File: tests/unit/oidc/test_views.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (di) [reply tests/unit/oidc/test_views.py:504]: Unless I'm missing something, this is passing on `main`, so changes shouldn't be necessary here.

File: warehouse/oidc/models/activestate.py (hunk 10 above)
Comment (th3coop) [reply warehouse/oidc/models/activestate.py:92]: The "ingredient" is the thing that is being built then published through our system.  It ultimately will be turned into the wheel that gets published.  I just checked with the team and it sounds like the ingredient name could potentially not match the PyPI project name but the wheel still be correct so I'm going to delete this comment.

File: warehouse/oidc/models/activestate.py (hunk 11 above)
Comment (th3coop) [reply warehouse/oidc/models/activestate.py:77]: I wasn't intending to, no.  In the form PR, I confirm the organization exists with an API call but I don't save anything: https://github.com/pypi/warehouse/pull/15168/files#diff-7d755d844461cdc27241084bdffddb5860d65107d1ac79d7a0e56c6b4c7d8f11R122

File: warehouse/oidc/models/activestate.py (hunk 12 above)
Comment (th3coop) [reply warehouse/oidc/models/activestate.py:123]: No, this property isn't necessary.

File: warehouse/oidc/models/activestate.py (hunk 13 above)
Comment (th3coop) [reply warehouse/oidc/models/activestate.py:138]: Agreed.  Done locally.

File: tests/unit/oidc/test_views.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (th3coop) [reply tests/unit/oidc/test_views.py:504]: Sorry, Dustin, I don't know what you mean.

File: warehouse/oidc/models/activestate.py (hunk 12 above)
Comment (th3coop) [reply warehouse/oidc/models/activestate.py:123]: I misspoke here, I do need this property.  With my understanding of how the `check_claim_invariant` will work, I need this propery.  The `builder` claim coming from ActiveState will always be `pypi-publisher`.  Am I using `check_claim_invariant` wrong?

File: tests/unit/oidc/test_views.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (di) [reply tests/unit/oidc/test_views.py:504]: Sorry, I mean that this test should have been passing without any changes, and I don't think any changes you need to introduce would cause it to fail, so these changes are probably unnecessary for this PR (although you're correct that this test is probably over-mocking, but we should probably fix that separately)

File: warehouse/oidc/models/activestate.py (hunk 11 above)
Comment (di) [reply warehouse/oidc/models/activestate.py:77]: I'm not familiar enough with Activestate: if a user deletes their organization named `foo`, can another user resurrect the same organization? I guess if we're also validating that the `actor_id` has not changed, even if that was possible this should be OK, but just want to validate my assumptions.

File: warehouse/oidc/models/activestate.py (hunk 12 above)
Comment (di) [reply warehouse/oidc/models/activestate.py:121]: Ah, you're correct, `check_claim_invariant` is using the value passed to the verifier _and_ the ground truth. This probably isn't actually necessary as the invariant value and the ground truth will always be the same, but changing that is definitely out of scope for this PR (cc @woodruffw for thoughts).

File: warehouse/migrations/versions/bd560a14a75c_.py (hunk 2 above)
Comment (di) [reply warehouse/migrations/versions/bd560a14a75c_.py:20]: FYI, looks like this migration is still here, we should remove it.

File: warehouse/oidc/models/github.py [hunk 26]
```diff
@@ -269,7 +269,6 @@ def reify(self, session):
         Returns a `GitHubPublisher` for this `PendingGitHubPublisher`,
         deleting the `PendingGitHubPublisher` in the process.
         """
-
```
Comment (di) [inline warehouse/oidc/models/github.py:272]: Stray whitespace removal here!

File: warehouse/oidc/models/activestate.py (hunk 11 above)
Comment (th3coop) [reply warehouse/oidc/models/activestate.py:77]: Ahh I understand and great point.  We don't allow re-use of deleted organization names due to possible name squatting.   And yeah, as you said, we're validating the user with the `actor_id`.

Comment (di) [reply warehouse/oidc/models/activestate.py:77]: Great, thanks for clarifying! Resolving this.

File: warehouse/oidc/models/activestate.py [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+from warehouse.oidc.interfaces import SignedClaims
+from warehouse.oidc.models._core import (
+    CheckClaimCallable,
+    OIDCPublisher,
+    PendingOIDCPublisher,
+)
+
+ACTIVESTATE_URL = "https://platform.activestate.com"
+
+ACTIVESTATE_OIDC_API_PATH = "api/v1/oauth/oidc"
```
Comment (woodruffw) [inline warehouse/oidc/models/activestate.py:34]: Nitpick: These look slightly duplicated against `utils.ACTIVESTATE_OIDC_ISSUER_URL`

File: warehouse/oidc/models/activestate.py [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+
+def _check_sub(
+    ground_truth: str, signed_claim: str, _all_signed_claims: SignedClaims
+) -> bool:
+    # We expect a string formatted as follows:
+    #  org:<orgName>:project:<projectName>
+    # where :OPTIONAL-STUFF is a concatenation of other job context
+    # metadata. We currently lack the ground context to verify that
+    # additional metadata, so we limit our verification to just the ORG/REPO
+    # component.
```
Comment (woodruffw) [inline warehouse/oidc/models/activestate.py:45]: Nitpick: This comment looks stale/slightly off: there's no mention of `:OPTIONAL-STUFF` 🙂

File: warehouse/oidc/models/activestate.py (hunk 27 above)
Comment (woodruffw) [reply warehouse/oidc/models/activestate.py:34]: (`ACTIVESTATE_OIDC_API_PATH` seems unused, so maybe it can be removed outright? And `ACTIVESTATE_URL` is used only within this module, so maybe it could be made into `_ACTIVESTATE_BASE_URL` or similar to emphasize that it's private.)

File: warehouse/oidc/models/github.py (hunk 26 above)
Comment (th3coop) [reply warehouse/oidc/models/github.py:272]: Doh!  I thought I got this already.  Fixing.

File: warehouse/oidc/models/activestate.py (hunk 27 above)
Comment (th3coop) [reply warehouse/oidc/models/activestate.py:34]: Good eye.  Changes incoming.

File: warehouse/migrations/versions/bd560a14a75c_.py (hunk 2 above)
Comment (th3coop) [reply warehouse/migrations/versions/bd560a14a75c_.py]: Yeah, i had backed out the change by accident. Just pushed a commit that straightens things out.

**Review by woodruffw — APPROVED** [review]
LGTM!

**Comment by th3coop** [comment]
yeah, let me know if you need anything else from me in this PR.  I'm currently doing some touch ups on #15168 based on this review.  It's almost ready.  I'll be pushing the fixes to the tests in a few minutes here I think.

**Review by di — APPROVED** [review]
LGTM! I'll wait to merge this until tomorrow AM.

**Comment by di** [comment]
@th3coop Actually, I don't have a button to bring this branch up-to-date with `main`, so I'll need you to rebase this for me to be able to merge it.

**Comment by th3coop** [comment]
No problem @di .  Done.

**Comment by woodruffw** [comment]
Looks like this needs one last branch update 🙂

**Comment by di** [comment]
I'll merge this in #15201 instead. 

@th3coop, you might want to enable this on future PRs: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork
